proposals/0145-implement-the-pipeline-operator.md

```diff
@@ -0,0 +1,447 @@
+
+if let imageURL = URL(string: imageURLString) {
+    if let imageData = NSData(contentsOf: imageURL) {
+        if let image = UIImage(data: imageData as Data) {
+            let imageView = UIImageView(image: image)
+        }
+    }
+}
+``` 
+
+Again, it is not very concise and makes it harder to read and understand the 
+code flow at first sight.
+
+As we can find in the great masterpiece [SICP, *Structure and Interpretation of Computer Programs*](https://mitpress.mit.edu/sicp/full-text/sicp/book/node4.html): 
+
+> *Computational processes* are abstract beings that inhabit computers. As they
+evolve, processes manipulate other abstract things called *data*. The evolution
+of a process is directed by a pattern of rules called a *program*. 
+**People create programs to direct processes**.
+
+We know that, in essence, programs transform data. Input data are somehow
+manipulated and returned as output data ─ so **it is very important to have have
+our have our data flow clearly defined and concise**.
+
+## Proposed solution
+
+Although Swift is not really a functional programming language, it does 
+support a few functional-programming patterns ─ and has a *well-designed 
+type system* ─, so many functional programming lovers ended up by 
+implementing their favorite functional idioms and libraries from other 
+programming languages in Swift. Good examples on this are libraries like
+[Swiftz](https://github.com/typelift/Swiftz),
+[Dollar](https://github.com/ankurp/Dollar)
+and even [some ports of Haskell's prelude](https://github.com/robrix/Prelude).
+
+A feature many developers are used to ─ and actually love ─ in many funcional
+programming languages is their **pipeline operator** ─ also known as 
+**pipe-forward** operator ─ `|>` ─, which helps to write readable functional 
+code by allowing developers to **apply the left hand side of the expression as 
+the first argument in the function on the right ─ enabling function calls to be chained together as successive operations**,
+in a more human readable way, in the correct flux of execution, reducing cognitive overhead.
+
+Taking a look at a few examples, our very first one ─ incrementing and 
+afterwards squaring a value, we would have something like this:
+
+```swift
+let myValue: Int = 66
+
+let myNewValue: Int = myValue
+                |> increment
+                |> square
+```
+
+In our grades data manipulation example, we would get something like:
+
+```swift
+let finalGrades = Database.allStudents()
+                  |> Grades.forStudents
+                  |> curve(average: 0.6)
+                  |> prepareGrades
+```
+
+Last but not least, our 3rd example ─ which involved *deep* optional chaining ─,
+after the **pipe-forward** operator goes like:
+
+```swift
+let imageURLString: String = "https://avatars2.githubusercontent.com/u/10639145"
+
+let imageView = imageURLString
+    |> { URL(string: $0) }
+    |> { NSData(contentsOf: $0 as URL) }
```

Comment:
Some notes here:
1. Should read `Data` instead of `NSData`
2. URL.init(string:) returns an optional, which cannot be pipelined to `Data.init(contentsOf:)` the bind operator (>>=) would need to be used here to follow the FP paradigms
3. If we're talking functional programming, then this could've been written more succinctly as:
```swift
let imageView = imageURLString |> URL.init(string:) |> UIImage.init(data:) |> UIImageView.init(image:)
```
Again, since there are optionals involved, this is not the best example for pipelining.